sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,934 @@
+                                       new_state);
+}
+
+dif_usbdev_result_t dif_usbdev_interface_enable(dif_usbdev_t *usbdev,
+                                                dif_usbdev_toggle_t new_state) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  if (!is_valid_toggle(new_state)) {
+    return kDifUsbdevResultGeneralError;
+  }
+
+  if (new_state == kDifUsbdevToggleEnable) {
+    mmio_region_nonatomic_set_bit32(
+        usbdev->base_addr, USBDEV_USBCTRL_REG_OFFSET, USBDEV_USBCTRL_ENABLE);
+  } else {
+    mmio_region_nonatomic_clear_bit32(
+        usbdev->base_addr, USBDEV_USBCTRL_REG_OFFSET, USBDEV_USBCTRL_ENABLE);
+  }
+
+  return kDifUsbdevResultOK;
+}
+
+dif_usbdev_rx_packet_get_info_result_t dif_usbdev_rx_packet_get_info(
+    dif_usbdev_t *usbdev, dif_usbdev_rx_packet_info_t *packet) {
+  if (usbdev == NULL || packet == NULL) {
+    return kDifUsbdevRxPacketGetInfoResultInvalidArgument;
+  }
+
+  // Check if there is already a read buffer operation in progress
+  if (buffer_op_is_reading(usbdev)) {
+    return kDifUsbdevRxPacketGetInfoResultBusy;
+  }
+
+  // Check if the RX FIFO is empty
+  if (mmio_region_get_bit32(usbdev->base_addr, USBDEV_USBSTAT_REG_OFFSET,
+                            USBDEV_USBSTAT_RX_EMPTY)) {
+    return kDifUsbdevRxPacketGetInfoResultNoPackets;
+  }
+
+  // Read fifo entry
+  const uint32_t fifo_entry =
+      mmio_region_read32(usbdev->base_addr, USBDEV_RXFIFO_REG_OFFSET);
+  // Fill packet info
+  packet->endpoint =
+      (fifo_entry >> USBDEV_RXFIFO_EP_OFFSET) & USBDEV_RXFIFO_EP_MASK;
+  packet->is_setup = fifo_entry & USBDEV_RXFIFO_SETUP;
+  packet->len =
+      (fifo_entry >> USBDEV_RXFIFO_SIZE_OFFSET) & USBDEV_RXFIFO_SIZE_MASK;
+  // Get the buffer entry that holds the payload
+  uint32_t buffer_entry =
+      (fifo_entry >> USBDEV_RXFIFO_BUFFER_OFFSET) & USBDEV_RXFIFO_BUFFER_MASK;
+  // Start a read buffer operation
+  buffer_op_start_read(usbdev);
+  usbdev->read_buffer_op_state =
+      (dif_usbdev_read_buffer_op_state_t){.buffer_entry = buffer_entry,
+                                          .remaining_bytes = packet->len,
+                                          .offset = 0};
+
+  return kDifUsbdevRxPacketGetInfoResultOK;
+}
+
+dif_usbdev_rx_packet_read_bytes_result_t dif_usbdev_rx_packet_read_bytes(
+    dif_usbdev_t *usbdev, uint8_t *dst, uint32_t dst_len,
+    uint32_t *bytes_written) {
+  if (usbdev == NULL || dst == NULL) {
+    return kDifUsbdevRxPacketReadBytesResultInvalidArgument;
+  }
+
+  // Make sure that a read buffer operation is in progress
+  if (!buffer_op_is_reading(usbdev)) {
+    return kDifUsbdevRxPacketReadBytesResultNotReading;
+  }
+  // bytes_to_copy is the minimum of remaining bytes and `dst_len`
+  uint32_t bytes_to_copy = usbdev->read_buffer_op_state.remaining_bytes;
+  if (bytes_to_copy > dst_len) {
+    bytes_to_copy = dst_len;
+  }
+  // Get the address of the buffer entry
+  uint32_t buffer_addr = USBDEV_BUFFER_REG_OFFSET +
```

Comment:
Consider factoring this out into a helper?

Reply:
Good point.